src/app/clusters/webrtc-transport-provider-server/webrtc-transport-provider-server.cpp

```diff
@@ -0,0 +1,546 @@
+    {
+        // If both video and audio stream IDs are missing, return CONSTRAINT_ERROR.
+        if (!videoStreamID.HasValue() && !audioStreamID.HasValue())
+        {
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+            return;
+        }
+
+        // Validate the stream usage according to resource management and stream priorities.
+        CHIP_ERROR err = mDelegate.ValidateStreamUsage(req.streamUsage, req.videoStreamID, req.audioStreamID);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(Zcl, "HandleProvideOffer: Invalid stream usage");
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ResourceExhausted);
+            return;
+        }
+
+        // Prepare delegate arguments.
+        Delegate::ProvideOfferRequestArgs args;
+        args.sessionId       = GenerateSessionId();
+        args.streamUsage     = req.streamUsage;
+        args.videoStreamId   = videoStreamID;
+        args.audioStreamId   = audioStreamID;
+        args.metadataOptions = req.metadataOptions;
+        args.peerNodeId      = peerNodeId;
+        args.sdp             = std::string(req.sdp.data(), req.sdp.size());
+
+        // Convert ICE servers list from DecodableList to vector.
+        if (req.ICEServers.HasValue())
+        {
+            auto iterator = req.ICEServers.Value().begin();
+            while (iterator.Next())
+            {
+                args.iceServers.Value().push_back(iterator.GetValue());
+            }
+        }
+
+        // Convert ICETransportPolicy from CharSpan to std::string.
+        if (req.ICETransportPolicy.HasValue())
+        {
+            args.iceTransportPolicy.SetValue(
+                std::string(req.ICETransportPolicy.Value().data(), req.ICETransportPolicy.Value().size()));
+        }
+
+        // Delegate processing: process the SDP offer, gather ICE candidates, create SDP answer, etc.
+        auto delegateStatus = Protocols::InteractionModel::ClusterStatusCode(mDelegate.HandleProvideOffer(args, outSession));
+        if (!delegateStatus.IsSuccess())
+        {
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, delegateStatus);
+            return;
+        }
+
+        // Store the new WebRTCSessionStruct in CurrentSessions.
+        UpsertSession(outSession);
+    }
+    else
+    {
+        // Re-offer: webRTCSessionID is not null.
+        uint16_t sessionId                    = webRTCSessionID.Value();
+        WebRTCSessionStruct * existingSession = FindSession(sessionId);
+
+        // If the session does not exist, respond with NOT_FOUND.
+        if (!existingSession)
+        {
+            ChipLogError(Zcl, "HandleProvideOffer: No existing session with ID=%u.", sessionId);
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::NotFound);
+            return;
+        }
+
+        // Verify that the session belongs to the accessing peer.
+        if (peerNodeId != existingSession->peerNodeID)
```

Comment:
Where is the fabric index check?

Note that WebRTCSessionStruct _does_ have a fabric index, since it's fabric-scoped.